R/driver-snowflake.R

```diff
@@ -87,3 +87,238 @@ setMethod("odbcDataType", "Snowflake",
+    # Connections to Snowflake are always over HTTPS.
+    port = 443
+  )
+  auth <- snowflake_auth_args(account, ...)
+  all <- utils::modifyList(c(args, auth), list(...))
+
+  # Respect the Snowflake Partner environment variable, if present.
+  if (is.null(all$application) && nchar(Sys.getenv("SF_PARTNER")) != 0) {
+    all$application <- Sys.getenv("SF_PARTNER")
+  }
+
+  arg_names <- tolower(names(all))
+  if (!"authenticator" %in% arg_names && !all(c("uid", "pwd") %in% arg_names)) {
+    warn(
+      c(
+        "x" = "Failed to detect ambient Snowflake credentials.",
+        "i" = "Supply `uid` and `pwd` to authenticate manually."
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+
+  all
+}
+
+# Returns a sensible driver name even if odbc.ini and odbcinst.ini do not
+# contain an entry for the Snowflake ODBC driver. For Linux and macOS we
+# default to known shared library paths used by the official installers.
+# On Windows we use the official driver name.
+snowflake_default_driver <- function() {
+  default_paths <- snowflake_default_driver_paths()
+  if (length(default_paths) > 0) {
+    return(default_paths[1])
+  }
+
+  fallbacks <- c("Snowflake", "SnowflakeDSIIDriver")
+  fallbacks <- intersect(fallbacks, odbcListDrivers()$name)
+  if (length(fallbacks) > 0) {
+    return(fallbacks[1])
+  }
+
+  abort(
+    c(
+      "Failed to automatically find Snowflake ODBC driver.",
+      i = "Set `driver` to known driver name or path."
+    ),
+    call = quote(DBI::dbConnect())
+  )
+}
+
+snowflake_default_driver_paths <- function() {
+  if (Sys.info()["sysname"] == "Linux") {
+    paths <- c(
+      "/opt/rstudio-drivers/snowflake/bin/lib/libsnowflakeodbc_sb64.so",
+      "/usr/lib/snowflake/odbc/lib/libSnowflake.so"
+    )
+  } else if (Sys.info()["sysname"] == "Darwin") {
+    paths <- "/opt/snowflake/snowflakeodbc/lib/universal/libSnowflake.dylib"
+  } else {
+    paths <- character()
+  }
+  paths[file.exists(paths)]
+}
+
+snowflake_server <- function(account) {
+  if (nchar(account) == 0) {
+    abort(
+      c(
+        "No Snowflake account ID provided.",
+        i = "Either supply `account` argument or set env var `SNOWFLAKE_ACCOUNT`."
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+  paste0(account, ".snowflakecomputing.com")
+}
+
+snowflake_auth_args <- function(account,
+                                uid = NULL,
+                                pwd = NULL,
+                                authenticator = NULL,
+                                ...) {
+  if (!is.null(uid) && !is.null(pwd)) {
+    return(list(uid = uid, pwd = pwd))
+  } else if (xor(is.null(uid), is.null(pwd))) {
+    abort(
+      c(
+        "Both `uid` and `pwd` must be specified to authenticate.",
+        i = "Or leave both unset to use ambient Snowflake credentials."
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+
+  # Manual override of the authentication method.
+  if (!is.null(authenticator)) {
+    list()
```

Comment:
```suggestion
    return(list())
```

As-is, this line doesn't early return. I believe the suggested is the intent?

Reply:
Done!